akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSubscriber.scala

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import akka.actor.{ ActorLogging, ActorRef, Deploy, Props }
+import akka.stream.actor.{ ActorSubscriberMessage, RequestStrategy }
+import akka.util.ByteString
+
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+object InputStreamSubscriber {
+  def props(completionPromise: Promise[Long], bufSize: Int): Props = {
+    require(bufSize > 0, "buffer size must be > 0")
+    Props(classOf[InputStreamSubscriber], completionPromise, bufSize).withDeploy(Deploy.local)
+  }
+
+  trait InputStreamSubscriberInMessage
+
+  case class Read(length: Int) extends InputStreamSubscriberInMessage
+
+  case object Close extends InputStreamSubscriberInMessage
+
+  trait InputStreamSubscriberOutMessage
+
+  case class Bytes(bytes: ByteString) extends InputStreamSubscriberOutMessage
+
+  case class Failure(msg: String) extends InputStreamSubscriberOutMessage
+
+}
+
+/** INTERNAL API */
+private[akka] class InputStreamSubscriber(bytesWrittenPromise: Promise[Long], bufSize: Int)
+  extends akka.stream.actor.ActorSubscriber
+  with ActorLogging {
+
+  var bytesRequestedDownstream: Option[(ActorRef, Int)] = None
+
+  var bytesSentDownstream: Long = 0
+
+  var onCompleteReceived: Boolean = false
+
+  var availableChunks: Vector[ByteString] = Vector.empty[ByteString]
+
+  override protected val requestStrategy = new RequestStrategy() {
+    override def requestDemand(remainingRequested: Int): Int = bytesRequestedDownstream match {
+      case Some(_) ⇒ if (remainingRequested > 0) 0 else 1
+      case None    ⇒ bufSize - availableChunks.size - remainingRequested
+    }
+  }
+
+  def receive = {
+    case ActorSubscriberMessage.OnNext(bytes: ByteString) ⇒
+      availableChunks :+= bytes
+      tryToReply(sendRest = false)
+
+    case ActorSubscriberMessage.OnError(cause) ⇒
+      tryToReply(sendRest = true)
+      bytesWrittenPromise.failure(cause)
+      context.stop(self)
```

Comment:
Yeah, this is the key point. I had a lot of troubles cause potentially any request from InputStream can be timed out because reactive stream already completed